File: isnetworkactive.go
package main

import (
	"bufio"
	"log/slog"
	"os"
	"strings"
)

func isNetworkActive() bool {
	// check ipv4 connections
	if hasNonLocalhostConnections("/proc/net/tcp") {
		return true
	}

	// check ipv6 connections
	return hasNonLocalhostConnections("/proc/net/tcp6")
}

func hasNonLocalhostConnections(procFile string) bool {
	file, err := os.Open(procFile)
	if err != nil {
		slog.Error("Could not open TCP connections file", slog.String("file", procFile), slog.Any("error", err))
		return true
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	scanner.Scan()

	for scanner.Scan() {
		fields := strings.Fields(scanner.Text())
		if len(fields) < 4 {
			continue
		}

		// Quick state check first - avoid expensive parsing if not established
		if !isStateEstablished(fields[3]) {
			continue
		}

		if !isLocalhostHex(fields[1]) {
			return true
		}
	}

	return false
}

func isStateEstablished(stateHex string) bool {
	// Quick check: established state is "01" in hex
	return stateHex == "01"
}

func isLocalhostHex(addrPort string) bool {
	colonIndex := strings.Index(addrPort, ":")
	if colonIndex == -1 {
		return false
	}

	hexAddr := addrPort[:colonIndex]

	// IPv4: 127.x.x.x range (127.0.0.0/8)
	if len(hexAddr) == 8 {
		return hexAddr[:2] == "7F"
	}

	// IPv6: Only ::1 exactly
	if len(hexAddr) == 32 {
		return hexAddr == "00000000000000000000000000000001"
	}

	return false
}

File: src/isnetworkactive.rs
use std::fs::File;
use std::{io, io::BufRead, io::BufReader};

pub fn is_network_active() -> Result<bool, io::Error> {
    Ok(
        has_non_localhost_connections("/proc/net/tcp")? ||
        has_non_localhost_connections("/proc/net/tcp6")?
    )
}

fn has_non_localhost_connections(proc_file_path: &str) -> Result<bool, io::Error> {
    let proc_file_handle = File::open(proc_file_path)?;
    let proc_file_reader = BufReader::new(proc_file_handle);

    for line in proc_file_reader.lines() {
        let line = line?;
        let fields: Vec<&str> = line.split_whitespace().collect();

        // if the line is too short or fields[3] (connection-state) is not "01" (established)
        if fields.len() < 4 || fields[3] != "01" {
            continue;
        }

        if !is_localhost_hex(fields[1]) {
            return Ok(true);
        }
    }

    Ok(false)
}

fn is_localhost_hex(address: &str) -> bool {
    let colon_index = address.find(":");
    let ip_hex_string = match colon_index {
        Some(index) => address.get(0..index),
        None => None,
    };

    match ip_hex_string {
        Some(hex) => {
            (hex.len() == 8 && hex.starts_with("7F")) ||
            (hex.len() == 32 && hex == "00000000000000000000000000000001")
        }
        None => false,
    }
}
